2009-11-24 Thomas Bernard <tbernard@proformatique.com>

	pf_atxfer_moh.patch
	$Revision$
	$Date$

	This patch fixes the bug of the waiting music being stopped during
	an atxfer when the transferer hang up before the target of the transfer
	pickup the phone.

	The patch was in fact done by Guillaume Knispel

Index: asterisk-1.4.26.2/res/res_features.c
===================================================================
--- asterisk-1.4.26.2.orig/res/res_features.c	2009-11-24 11:22:53.000000000 +0100
+++ asterisk-1.4.26.2/res/res_features.c	2009-11-24 11:26:19.000000000 +0100
@@ -1032,23 +1032,55 @@
 	                               newchan->name,
 	                               newchan->uniqueid);
 
-	if (check_compat(transferer, newchan)) {
-		/* we do mean transferee here, NOT transferer */
-		finishup(transferee);
-		return -1;
-	}
-	memset(&bconfig,0,sizeof(struct ast_bridge_config));
-	ast_set_flag(&(bconfig.features_caller), AST_FEATURE_DISCONNECT);
-	ast_set_flag(&(bconfig.features_callee), AST_FEATURE_DISCONNECT);
-	res = ast_bridge_call(transferer, newchan, &bconfig);
-	ast_log(LOG_DEBUG, "do_atxfer ast_bridge_call(%s, %s, ...) returned %d\n", transferer->name, newchan->name, res);
-	if (newchan->_softhangup || !transferer->_softhangup) {
-		ast_hangup(newchan);
-		if (ast_stream_and_wait(transferer, xfersound, transferer->language, ""))
-			ast_log(LOG_WARNING, "Failed to play transfer sound!\n");
-		finishup(transferee);
-		transferer->_softhangup = 0;
-		return FEATURE_RETURN_SUCCESS;
+	if (!ast_check_hangup(transferer)) {
+		if (check_compat(transferer, newchan)) {
+			/* we do mean transferee here, NOT transferer */
+			finishup(transferee);
+			return -1;
+		}
+		memset(&bconfig,0,sizeof(struct ast_bridge_config));
+		ast_set_flag(&(bconfig.features_caller), AST_FEATURE_DISCONNECT);
+		ast_set_flag(&(bconfig.features_callee), AST_FEATURE_DISCONNECT);
+		res = ast_bridge_call(transferer, newchan, &bconfig);
+		if (newchan->_softhangup || !transferer->_softhangup) {
+			ast_hangup(newchan);
+			if (ast_stream_and_wait(transferer, xfersound, transferer->language, ""))
+				ast_log(LOG_WARNING, "Failed to play transfer sound!\n");
+			finishup(transferee);
+			transferer->_softhangup = 0;
+			return FEATURE_RETURN_SUCCESS;
+		}
+	} else {
+		int connected;
+		switch ((int)newchan->_state) {
+		case AST_STATE_DIALING:
+		case AST_STATE_RING:
+		case AST_STATE_RINGING:
+			connected = 0;
+			while (!connected && (ast_waitfor(newchan, -1) >= 0)) {
+				f = ast_read(newchan);
+				if (f == NULL)
+					break;
+				if (f->frametype == AST_FRAME_CONTROL
+				    && f->subclass == AST_CONTROL_ANSWER)
+					connected = 1;
+				ast_frfree(f);
+			}
+			if (!connected) {
+				ast_hangup(newchan);
+				finishup(transferee);
+				return -1;
+			}
+			/* fall through */
+
+		case AST_STATE_UP:
+			break;
+
+		default:
+			ast_hangup(newchan);
+			finishup(transferee);
+			return FEATURE_RETURN_SUCCESS;
+		}
 	}
 
 	if (check_compat(transferee, newchan)) {
